contract/src/main/kotlin/io/provenance/scope/loan/utility/FundingDataValidation.kt

```diff
@@ -0,0 +1,149 @@
+                (fundingStatus.status in listOf("UNFUNDED", "INITIATED", "FUNDED", "CANCELLED"))
+                    orError "Funding status must be valid",
+                fundingStatus.effectiveTime.isValidFundingTime()
+                    orError "Funding status must have valid effective time",
+            )
+            when (fundingStatus.status) {
+                "FUNDED" -> {
+                    requireThat(
+                        funding.started.isValidFundingTime()   orError "Completed funding's start time must be valid",
+                        funding.completed.isValidFundingTime() orError "Completed funding's end time must be valid",
+                    )
+                    if (funding.started.isValidFundingTime() && funding.completed.isValidFundingTime()) {
+                        requireThat(
+                            (funding.completed.toInstant() >= funding.started.toInstant()) orError "Funding end time must be after start time",
+                        )
+                    }
+                }
+                "CANCELLED" -> {
+                    requireThat(
+                        funding.completed.isValidFundingTime() orError "Cancelled funding's end time must be valid",
+                    )
+                }
+                !in listOf("UNFUNDED", "UNKNOWN") -> {
+                    requireThat(
+                        funding.started.isValidFundingTime() orError "Funding start time must be valid",
+                    )
+                }
+            }
+        } ?: raiseError("Funding status must be set")
+        if (funding.disbursementsCount > 0) {
+            val incomingDisbursementIds = mutableMapOf<String, UInt>()
+            funding.disbursementsList.requireThatEach { disbursement ->
+                disbursement.takeIf { it.isSet() }?.also { setDisbursement ->
+                    setDisbursement.id.takeIf { it.isSet() }?.value?.also { disbursementId ->
+                        incomingDisbursementIds[disbursementId] = incomingDisbursementIds.getOrDefault(disbursementId, 0U) + 1U
+                    }
+                    requireThat(
+                        setDisbursement.id.isValid() orError "Disbursement must have valid ID",
+                        (setDisbursement.amount.value.toDoubleOrNull()?.let { it >= 0 } ?: true)
+                            orError "Disbursement amount must not be negative",
+                    )
+                    setDisbursement.status.takeIf { it.isSet() }?.also { disbursementStatus ->
+                        requireThat(
+                            disbursementStatus.status.isNotBlank()                orError "Disbursement status must not be empty",
+                            disbursementStatus.effectiveTime.isValidFundingTime() orError "Disbursement must have valid effective time",
+                        )
+                        when (disbursementStatus.status) {
+                            in listOf("COMPLETED", "FUNDED") -> {
```

Comment:
Did we confirm if Disbursements need a COMPLETED status? Currently they just equal the Funding Status states.

Reply:
So I didn't confirm that, but I would be comfortable leaving this be because only the funding status string is being validated to be one of the four values you mentioned, not the disbursement status, so I see no harm in validating "if a disbursement status is `COMPLETED`, it should have logical values for `started` and `completed`.

Reply:
I'm good with leaving it. Hopefully shouldn't need it but it won't hurt.